runtime/lua/vim/lsp/util.lua

```diff
@@ -525,26 +525,57 @@ function M.make_floating_popup_options(width, height, opts)
   }
 end
 
+local function _add_tagstack(item)
+  local function should_add(new_item)
+    local stack = vim.fn.gettagstack()
+
+    -- Check if we're at the bottom of the tagstack.
+    if stack.curidx <= 1 then return true end
+
+    local top_item = stack.items[stack.curidx-1]
+
+    -- Check if the item at the top of the tagstack is exactly the
+    -- same as the one we want to push.
+    if top_item.tagname ~= new_item.tagname then return true end
+    for i, v in ipairs(top_item.from) do
+      if v ~= new_item.from[i] then return true end
+    end
+    return false
+  end
+
+  -- Check whether item is not already at the top of the tagstack.
+  if not should_add(item) then return end
+  local winid = vim.fn.win_getid()
+  local items = {item}
+  vim.fn.settagstack(winid, {items=items}, 't')
+end
+
 function M.jump_to_location(location)
   -- location may be Location or LocationLink
   local uri = location.uri or location.targetUri
   if uri == nil then return end
-  local bufnr = vim.uri_to_bufnr(uri)
-  -- Save position in jumplist
-  vim.cmd "normal! m'"
-
-  -- Push a new item into tagstack
-  local from = {vim.fn.bufnr('%'), vim.fn.line('.'), vim.fn.col('.'), 0}
-  local items = {{tagname=vim.fn.expand('<cword>'), from=from}}
-  vim.fn.settagstack(vim.fn.win_getid(), {items=items}, 't')
 
   --- Jump to new location (adjusting for UTF-16 encoding of characters)
+  local bufnr = vim.uri_to_bufnr(uri)
   api.nvim_set_current_buf(bufnr)
```

Comment:
It seems weird that this jumps to the new buffer before checking where it came from.

Am I wrong there?

Reply:
UX-wise, it should act as a no-op. The tagstack check is done afterwards. I'm not sure about any possible overheads involving it though.

Reply:
Sorry, @tjdevries, you were totally right about this. This was not working, actually, when jumping from one buffer to another. Thank you very much for hinting me about this.

I have redone everything, including tests. They are now documented. :cat:

Reply:
<3 That's what I was wondering. Thanks for following up on this!